File: java/src/com/chargebee/samples/StripeJsCheckout.java
package com.chargebee.samples;

import com.chargebee.APIException;
import com.chargebee.Environment;
import com.chargebee.Result;
import com.chargebee.exceptions.InvalidRequestException;
import com.chargebee.exceptions.PaymentException;
import com.chargebee.models.Address;
import com.chargebee.models.Customer;
import com.chargebee.models.Subscription;
import java.io.IOException;
import java.io.PrintWriter;
import java.net.*;
import javax.servlet.ServletException;
import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import static com.chargebee.samples.common.Utils.*;
import static com.chargebee.samples.common.ErrorHandler.*;

/**
 * Demo on how to create subscription with ChargeBee API using stripe temporary token and
 * adding shipping address to the subscription for shipping of product.
 */
public class StripeJsCheckout extends HttpServlet {

    /**
     * Processes HTTP request
     * <code>POST</code> method, creates the subscription and also adds address
     * to the subscription
     *
     * @param request servlet request
     * @param response servlet response
     * @throws ServletException if a servlet-specific error occurs
     * @throws IOException if an I/O error occurs
     */
    protected void processRequest(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        /*
         * Setting the Content-Type as application/json.
         */
        response.setHeader("Content-Type", "application/json;charset=utf-8");
        PrintWriter out = response.getWriter();
        
        validateParameters(request);
        try {
            Result result = createSubscription(request);
            
            addShippingAddress(request, result.subscription().id(), result.customer());
            
            /* Forwarding to thank you page after successful create subscription.
             */
            //Writing json. Suggestion: Use proper json library
            out.write("{\"forward\": \"/stripe_js/thankyou.html\"}");
            
        } catch(PaymentException e) {
            handleTempTokenErrors(e, response, out);
        } catch(InvalidRequestException e){
            handleInvalidRequestErrors(e, response, out, "plan_id");
        } catch(Exception e) {
            handleGeneralErrors(e, response, out);
        } finally {
            out.close();
        }
    }
    
    @Override
    protected void doPost(HttpServletRequest request, HttpServletResponse response)
            throws ServletException, IOException {
        processRequest(request, response);
    }

    @Override
    public String getServletInfo() {
        return "Demo on how to create subscription with the ChargeBee API using stripe temporary token";
    }

    /**
     * Adds the shipping address to an existing subscription. The first name
     * & the last name for the shipping address is get from the customer 
     * account information in ChargeBee.
     *
     * @param req the HttpServletRequest from the client
     * @param subscriptionId Identifier of the subscription to attach the shipping address to.
     * @param customer Customer
     * @throws IOException
     */
    public void addShippingAddress(HttpServletRequest req, String subscriptionId, Customer customer) 
            throws IOException, Exception {
        /* 
         * Adding address to the subscription for shipping product to the customer.
         * Sends request to the ChargeBee server and adds the shipping address 
         * for the given subscription Id.
         */
        Result result = Address.update()
                .label("shipping_address")
                .subscriptionId(subscriptionId)
                .firstName(customer.firstName())
                .lastName(customer.lastName())
                .addr(req.getParameter("addr"))
                .extendedAddr(req.getParameter("extended_addr"))
                .city(req.getParameter("city"))
                .state(req.getParameter("state"))
                .zip(req.getParameter("zip_code")).request();

    }

    /**
     * Creates the subscription in ChargeBee using the checkout details and 
     * stripe temporary token provided by stripe.
     *
     * @param req HttpServletRequest
     * @return Result
     * @throws IOException
     */
    public Result createSubscription(HttpServletRequest req) 
            throws IOException, Exception {
        
        /* 
         * planId is id of the plan present in the ChargeBee app. 
         * For demo purpose a plan with id 'annual' is hard coded and should exist 
         * at your ChargeBee site. It can be also be received from request by 
         * making customer selecting the plan of their choice at client side.
         */
        String planId = "annual";
 
        /* Sends request to the ChargeBee server to create the subscription from 
         * the parameters received. The result will have subscription attributes, 
         * customer attributes and card attributes.
         */
        Result result = Subscription.create()
                .planId(planId)
                .customerEmail(req.getParameter("customer[email]"))
                .customerFirstName(req.getParameter("customer[first_name]"))
                .customerLastName(req.getParameter("customer[last_name]"))
                .customerPhone(req.getParameter("customer[phone]"))
                .cardTmpToken(req.getParameter("stripeToken")).request();
        
        return result;
    }
}

File: php/php_src/StripeJsCheckout.php
<?php
/*
 * Adding ChargeBee php librariesand configuration files.
 */
require_once(dirname(__FILE__) . "/Config.php");
require_once(dirname(__FILE__) . "/ErrorHandler.php");
require_once(dirname(__FILE__) . "/Util.php");

if ($_POST) {
	validateParameters($_POST);
    try {
        $result = createSubscription();
        addShippingAddress($result->subscription(), $result->customer());
        $jsonResp = array();
        
        /*
         * Forwarding to success page after successful create subscription in ChargeBee.
         */
        $queryParameters = "name=" . urlencode($result->customer()->firstName) 
                            . "&planId=" . urlencode($result->subscription()->planId);        
        $jsonResp["forward"] = "thankyou.html";
        echo json_encode($jsonResp, true);
        
    } catch(ChargeBee_PaymentException $e) {
   	 	handleTempTokenErrors($e);
    } catch(ChargeBee_InvalidRequestException $e) {
		handleInvalidRequestErrors($e, "plan_id");
    } catch(Exception $e) {
		handleGeneralErrors($e);
    }
}


/* Creates the subscription in ChargeBee using the checkout details and 
 * stripe temporary token provided by stripe.
 */
function createSubscription() {
    
    /*
     * Constructing a parameter array for create subscription api. 
     * It will have account information, the temporary token got from Stripe and
     * plan details.
     * For demo purpose a plan with id 'annual' is hard coded.
     * Other params are obtained from request object.
     * Note : Here customer object received from client side is sent directly 
     *        to ChargeBee.It is possible as the html form's input names are 
     *        in the format customer[<attribute name>] eg: customer[first_name] 
     *        and hence the $_POST["customer"] returns an associative array of the attributes.
     *               
     */
    $createSubscriptionParams = array(
        "planId" => "basic",
        "customer" => $_POST['customer'],
        "card" => array(
            "tmp_token" => $_POST['stripeToken']
    ));

    /* 
    * Sending request to the chargebee server to create the subscription from 
    * the parameters received. The result will have customer,subscription and 
    * card attributes.
    */
    $result = ChargeBee_Subscription::create($createSubscriptionParams);
    
    return $result;
}
/*
 * Adds the shipping address to an existing subscription. The first name
 * & the last name for the shipping address is got from the customer 
 * account information.
 */
function addShippingAddress($subscription, $customer) {
   /* 
    * Adding address to the subscription for shipping product to the customer.
    * Sends request to the ChargeBee server and adds the shipping address 
    * for the given subscription Id.
    */
    $result = ChargeBee_Address::update(array(
                "subscription_id" => $subscription->id,
                "label" => "shipping_address",
                "first_name" => $customer->firstName,
                "last_name" => $customer->lastName,
                "addr" => $_POST['addr'],
                "extended_addr" => $_POST['extended_addr'],
                "city" => $_POST['city'],
                "state" => $_POST['state'],
                "zip" => $_POST['zip_code']
    ));
    $address = $result->address();
    return $address;
}
?>

